File: 4-langs/tree.py
import attr


@attr.s
class Node:
    value = attr.ib()
    left = attr.ib()
    right = attr.ib()

    def insert(self, value):
        attr = 'left' if value < self.value else 'right'
        obj = getattr(self, attr)
        node = obj.insert(value) if obj else Node(value, None, None)
        setattr(self, attr, node)
        return self


def sum_tree(tree):
    if tree is None:
        return 0
    return tree.value + sum_tree(tree.left) + sum_tree(tree.right)


if __name__ == '__main__':
    from random import shuffle, seed

    seed(42)
    tree = Node(0, None, None)
    values = list(range(1, 10))
    shuffle(values)
    for val in values:
        tree.insert(val)

    print(sum_tree(tree))

File: pyweb/go/tree.go
package main

import (
	"fmt"
)

type Tree struct {
	value       int
	left, right *Tree
}

func walk(tree *Tree, ch chan int) {
	if tree == nil {
		return
	}
	ch <- tree.value
	walk(tree.left, ch)
	walk(tree.right, ch)
}

func leaves(tree *Tree) chan int {
	ch := make(chan int)
	go func() {
		walk(tree, ch)
		close(ch)
	}()
	return ch
}

func main() {
	tree := &Tree{1, &Tree{2, nil, nil}, &Tree{3, nil, nil}}
	for v := range leaves(tree) {
		fmt.Println(v)
	}
}
